### PR TITLE
Graceful shutdown / restart / interupt

### PR DESCRIPTION
**Pull request by ArcasCZ**
This PR let's the bot gracefuly shutdown / restart when:

- using !strawberry shutdown
- using !strawberry restart
- receiving SIGKILL / SIGTERM (docker container stop, CTRL+C, ...)

### CODE REVIEW

File: pie/bot.py [hunk 1]
```diff
@@ -58,6 +72,11 @@ async def update_app_info(self):
         else:
             self.owner_ids = {app.owner.id}
 
+    async def close(self, signal: str = None) -> None:
+        message = f"Signal{signal} received. " if signal else ""
```
Comment (musakai) [inline pie/bot.py:76]: No space is here required? I would use something like 

`message = f"Signal '{signal}' received. " if signal else "Unknown"`

Comment (ArcasCZ) [reply pie/bot.py:76]: The `close()` method is not used only for the signal handling, but also for shutting down the bot from !strawberry restart / shutdown commands. In those cases, the message can't contain signal. However, I might have an idea how to handle this better way.

Comment (ArcasCZ) [reply pie/bot.py]: Please check latest commit. It should be addressed

File: modules/base/admin/module.py [hunk 2]
```diff
@@ -645,5 +646,5 @@ async def spamchannel_primary(self, ctx, channel: discord.TextChannel):
         )
 
 
-async def setup(bot) -> None:
+async def setup(bot: Strawberry) -> None:
```
Comment (musakai) [inline modules/base/admin/module.py:649]: Documentation

File: pie/bot.py [hunk 3]
```diff
@@ -50,6 +53,17 @@ def __init__(self):
         self.bot_log = logger.Bot.logger(self)
         self.guild_log = logger.Guild.logger(self)
 
+    async def setup_hook(self):
```
Comment (musakai) [inline pie/bot.py:56]: Documentation

File: pie/bot.py [hunk 4]
```diff
@@ -58,6 +72,11 @@ async def update_app_info(self):
         else:
             self.owner_ids = {app.owner.id}
 
+    async def close(self, signal: str = None) -> None:
```
Comment (musakai) [inline pie/bot.py]: Documentation

File: pie/bot.py [hunk 5]
```diff
@@ -50,6 +53,17 @@ def __init__(self):
...
 
+    async def setup_hook(self):
+        self.loop.add_signal_handler(
+            signal.SIGINT, partial(self.handle_signal, "SIGINT")
+        )
+        self.loop.add_signal_handler(
+            signal.SIGTERM, partial(self.handle_signal, "SIGTERM")
+        )
+
+    def handle_signal(self, signal):
```
Comment (musakai) [inline pie/bot.py:64]: Documentation

File: pie/bot.py (hunk 1 above)
Comment (musakai) [reply pie/bot.py]: LGTM

File: modules/base/admin/module.py (hunk 2 above)
Comment (ArcasCZ) [reply modules/base/admin/module.py:649]: I think here it's not really necessary, see: https://discordpy.readthedocs.io/en/latest/ext/commands/api.html?highlight=load_extension#discord.ext.commands.Bot.load_extension

File: pie/bot.py (hunk 3 above)
Comment (ArcasCZ) [reply pie/bot.py:56]: Added

File: pie/bot.py (hunk 4 above)
Comment (ArcasCZ) [reply pie/bot.py]: Added

File: pie/bot.py (hunk 5 above)
Comment (ArcasCZ) [reply pie/bot.py:64]: Done

Comment (musakai) [reply pie/bot.py:66]: LGTM

**Review by musakai — APPROVED** [review]
LGTM

**Pull request merged**